### PR TITLE
Bolt

### PR DESCRIPTION
**Pull request by cwharris**
+ .gitignore: added pycharm, vscode related directories
+ bolt / cypher support + jupyter notebook demo

### CODE REVIEW

**Comment by cwharris** [comment]
I still need to run through this to clean up.

@lmeyerov i know we talked about performance / choking while converting data, but Bolt’s graph’s relationships has a structure such as `{is:, from:, to:, props:{...}}`, and I’ve not figured out how to view those props in the graph viewer without collapsing them to a single dictionary.

File: graphistry/plotter.py [hunk 1]
```diff
@@ -547,7 +549,6 @@ def bind(df, pbname, attrib, default=None):
             elif default:
                 df[pbname] = df[default]
 
-
```
Comment (cwharris) [inline graphistry/plotter.py]: Shouldn’t have removed this line

**Comment by cwharris** [comment]
Ready for review.

File: demos/Neo4jBoltCypher.ipynb [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "GRAPHISTRY = {\n",
+    "    'server': 'labs.graphistry.com',\n",
+    "    'api': 2,\n",
+    "    'key': '1e8e6951d29252a24069232813ca134ee66dfc25ef6846e7d4ff7e434eaed14b'\n",
```
Comment (lmeyerov) [inline demos/Neo4jBoltCypher.ipynb]: oopsies

File: graphistry/pygraphistry.py [hunk 3]
```diff
@@ -194,6 +209,7 @@ def register(key=None, server=None, protocol=None, api=None, certificate_validat
         PyGraphistry.protocol(protocol)
         PyGraphistry.certificate_validation(certificate_validation)
         PyGraphistry.authenticate()
+        PyGraphistry.set_bolt_driver(bolt)
```
Comment (lmeyerov) [inline graphistry/pygraphistry.py]: Should there be a `None` check here?

File: graphistry/plotter.py [hunk 4]
```diff
@@ -653,3 +653,49 @@ def _make_vgraph_dataset(self, edges, nodes, name):
...
+            res._edges = edges
+        return res.bind(\
+            edge_title='_bolt_relationship_id',\
+            source='_bolt_source_id',\
+            destination='_bolt_destination_id'
+        )
+
+
+    @staticmethod
+    def _bolt_graph_to_dataframe(graph):
```
Comment (lmeyerov) [inline graphistry/plotter.py]: Can we factor this stuff out into another file?

File: graphistry/plotter.py [hunk 5]
```diff
@@ -653,3 +653,49 @@ def _make_vgraph_dataset(self, edges, nodes, name):
         dataset = vgraph.create(elist, filtered_nlist, sources, dests, nodeid, node_map, name)
         dataset['encodings'] = encodings
         return dataset
+
+
+    def bolt(self, driver):
```
Comment (lmeyerov) [inline graphistry/plotter.py]: If you put python comments here, then a notebook user can do `help(g.bolt)` to learn about how to use this

File: graphistry/plotter.py [hunk 6]
```diff
@@ -653,3 +653,49 @@ def _make_vgraph_dataset(self, edges, nodes, name):
...
+        from neo4j import GraphDatabase, Driver
+        if driver is not None and isinstance(driver, Driver) == False:
+            driver = GraphDatabase.driver(**driver)
+        res = copy.copy(self)
+        res._bolt_driver = driver
+        return res
+
+
+    def cypher(self, query, params={}):
```
Comment (lmeyerov) [inline graphistry/plotter.py]: If you put python comments here, then a notebook user can do `help(g.cypher)` to learn about how to use this

File: graphistry/plotter.py [hunk 7]
```diff
@@ -164,6 +164,7 @@ def bind(self, source=None, destination=None, node=None,
         res._point_label = point_label or self._point_label
         res._point_color = point_color or self._point_color
         res._point_size = point_size or self._point_size
+
```
Comment (lmeyerov) [inline graphistry/plotter.py]: Just as a habit, nice not to have unrelated formatting changes as part of a PR...

File: graphistry/plotter.py [hunk 8]
```diff
@@ -653,3 +653,49 @@ def _make_vgraph_dataset(self, edges, nodes, name):
...
+        return res
+
+
+    def cypher(self, query, params={}):
+        res = copy.copy(self)
+        driver = self._bolt_driver or PyGraphistry._config['bolt_driver']
+        with driver.session() as session:
+            graph = session.run(query, **params).graph()
+            edges = Plotter._bolt_graph_to_dataframe(graph)
+            res._edges = edges
```
Comment (lmeyerov) [inline graphistry/plotter.py]: You seem to be missing the nodes, which matters if they have their own attributes

File: graphistry/plotter.py [hunk 9]
```diff
@@ -653,3 +653,49 @@ def _make_vgraph_dataset(self, edges, nodes, name):
...
+        return res.bind(\
+            edge_title='_bolt_relationship_id',\
```
Comment (lmeyerov) [inline graphistry/plotter.py]: Do we know edge_title is the prettified edge name? may be better to leave unspecified and let graphistry figure something out..

File: graphistry/plotter.py (hunk 8 above)
Comment (lmeyerov) [reply graphistry/plotter.py]: In the original code I sent, there should have been something like:

if nodes:
   g = g.bind(node='nodeidcolumn').nodes(nodes_df)

File: graphistry/plotter.py [hunk 10]
```diff
@@ -653,3 +653,49 @@ def _make_vgraph_dataset(self, edges, nodes, name):
...
+            source='_bolt_source_id',\
+            destination='_bolt_destination_id'
+        )
+
+
+    @staticmethod
+    def _bolt_graph_to_dataframe(graph):
+        import pandas as pd
+        return pd.DataFrame([
+            Plotter.merge_two_dicts(
```
Comment (lmeyerov) [inline graphistry/plotter.py]: FWIW, probably worth testing we can do a 500K node / 2M edge graph this way, because our users will want that. this function call sticks out wrt that.

File: graphistry/pygraphistry.py [hunk 11]
```diff
@@ -152,8 +158,17 @@ def certificate_validation(value=None):
             requests.packages.urllib3.disable_warnings()
         PyGraphistry._config['certificate_validation'] = v
 
+
+    @staticmethod
+    def set_bolt_driver(driver=None):
+        from neo4j import GraphDatabase, Driver
+        if driver is not None and isinstance(driver, Driver) == False:
+            driver = GraphDatabase.driver(**driver)
+        PyGraphistry._config['bolt_driver'] = driver
```
Comment (lmeyerov) [inline graphistry/pygraphistry.py]: code clone w/ `bolt()`?

File: graphistry/pygraphistry.py (hunk 3 above)
Comment (cwharris) [reply graphistry/pygraphistry.py]: I added a null check to `.pygraphistry.bolt_util`

File: graphistry/plotter.py (hunk 4 above)
Comment (cwharris) [reply graphistry/plotter.py]: Moved to `.pygraphistry.bolt_util`.

File: graphistry/plotter.py (hunk 8 above)
Comment (cwharris) [reply graphistry/plotter.py]: :+1:

File: graphistry/plotter.py (hunk 9 above)
Comment (cwharris) [reply graphistry/plotter.py]: Leaving it unset and letting the user handle it.

File: graphistry/plotter.py (hunk 10 above)
Comment (cwharris) [reply graphistry/plotter.py]: As it stands, this works with the entire silk road data set from `bolt://violet-russel-mills-natalia.graphstory.cloud:7687`.

File: graphistry/pygraphistry.py (hunk 11 above)
Comment (cwharris) [reply graphistry/pygraphistry.py]: consolidated in `bolt_util`

File: graphistry/plotter.py (hunk 5 above)
Comment (cwharris) [reply graphistry/plotter.py]: added some comments. feel free to update them with something more user friendly. I'm not sure how to appease data scientists with these comments.

File: graphistry/plotter.py (hunk 6 above)
Comment (cwharris) [reply graphistry/plotter.py]: :+1:

File: graphistry/bolt_util.py [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    try:
+        from neo4j import GraphDatabase, Driver
+        if isinstance(driver, Driver):
+            return driver
+        return GraphDatabase.driver(**driver)
+    except ImportError:
+        return None
+
+
+def bolt_graph_to_dataframe(graph):
```
Comment (lmeyerov) [inline graphistry/bolt_util.py]: Rename `bolt_graph_to_edges` or`bolt_graph_to_edges_dataframe` ? (Inconsistent w/ bolt_graph_to_nodes)

File: graphistry/plotter.py [hunk 13]
```diff
@@ -653,3 +655,27 @@ def _make_vgraph_dataset(self, edges, nodes, name):
...
+
+
+    def cypher(self, query, params={}):
+        res = copy.copy(self)
+        driver = self._bolt_driver or PyGraphistry._config['bolt_driver']
+        with driver.session() as session:
+            bolt_statement = session.run(query, **params)
+            graph = bolt_statement.graph()
+            edges = bolt_util.bolt_graph_to_dataframe(graph)
+            nodes = bolt_util.bolt_graph_to_nodes(graph)
```
Comment (lmeyerov) [inline graphistry/plotter.py]: Are we guaranteed to have nodes, e.g., what happens if someone does cypher ".... return r" instead of ".... return a, r, b"? (Not sure what `.graph()` does here..)

File: graphistry/pygraphistry.py [hunk 14]
```diff
@@ -250,7 +264,31 @@ def hypergraph(raw_events, entity_types=None, opts={}, drop_na=True, drop_edge_a
         return hyper.Hypergraph().hypergraph(PyGraphistry, raw_events, entity_types, opts, drop_na, drop_edge_attrs, verbose)
 
 
+    @staticmethod
+    def bolt(driver = None):
+        """
+
+        :param driver: Neo4j Driver or arguments for GraphDatabase.driver(**{...})**
```
Comment (lmeyerov) [inline graphistry/pygraphistry.py]: Put example call?

File: graphistry/plotter.py (hunk 13 above)
Comment (cwharris) [reply graphistry/plotter.py]: It works implicitly. No nodes = no node info, but since we have edges, the graph is still visualized.
